### DISCUSSION TITLE
Profile 8.4

### DISCUSSION THREAD

**Discussion by joaoserra** [quietvoid/dovi_tool#206] [Q&A] [answered]
Hi, 

This issue is not about dovi_tool, mas I'm hoping you can help me with one question.

I'm trying to understand how profile 8.4 works. As far as I know, this is only used by the iPhone. (at least, so far)
If I understood it correctly, profile 8.4 is HLG+RPU. But HLG is a scene-referred system and not based on absolute values like PQ, so what's the point of metadata?

So what will a Display do with a HLG bitstream and Dolby Metadata?Maybe the Display converts the HLG to PQ with a 1000nits peak (using the algorithm described in in https://www.bbc.co.uk/rd/sites/50335ff370b5c262af000004/assets/592eea8006d63e5e5200f90d/BBC_HDRTV_PQ_HLG_Transcode_v2.pdf / BT.2390) and then apply the rpu metadata? 
But what metadata levels does profile 8.4 contain? Maybe only basic metadata like L1? I'm not sure the rest of the metadata levels can be calculated in real-time video recording.
Thanks!

**Comment by quietvoid** [comment, marked as answer]
If I remember correctly the static reshaping from iPhone changes the image to PQ 10000 nits.
It then applies the metadata in the same way as if it was already a PQ base video.

On Apple devices, for some reason Apple is limiting the metadata levels being used.
On older versions, there was only L1.

Now I think the recording or encoding calculates some metadata automatically, so there are many more levels.

**Reply by joaoserra** [reply]
Thanks!

When you say "iPhone changes the image to PQ 10000 nits", it's when the iPhone are playing the image, not capturing/encoding, right? HLG to PQ is not done at play-time. (i.e. The bitstream on the file is HLG)

I don't have an iPhone 12, but just downloaded an iPhone DV 8.4 sample from the internet and it only has LV1 on the cmv29_metadata.

Maybe this is from an older iOS version? (I'm saying this because you mention that "LV1 only" was an older versions.

Another question: Is there any public documentation on the format of the RPU? I can understand the cmv29_metadata block because is documented on CTA-861 and SMPTE ST 2094-10, but i don't find documentation for the rest of the RPU.

**Reply by quietvoid** [reply]
See saindriches' comment, it is more much more indepth than I could explain it.

> Another question: Is there any public documentation on the format of the RPU? I can understand the cmv29_metadata block because is documented on CTA-861 and SMPTE ST 2094-10, but i don't find documentation for the rest of the RPU.

The only documentation is the code in the `dolby_vision` crate.
Otherwise there are only patents and some incomplete specifications.

RPU: https://patents.google.com/patent/US10701399B2/en
Display mapping (CM v2.9): https://www.etsi.org/deliver/etsi_gs/CCM/001_099/001/01.01.01_60/gs_ccm001v010101p.pdf

**Comment by saindriches** [comment]
**TL;DR**
Dolby Vision Display Management is done in IPT color space, so HLG needs to be converted.
There are two versions of Apple's Profile 8.4, the early version is CMv2.9, and only contains L1.

In general, Dolby Vision content can be produced in two ways: encode from a master, or skip the encoding step (for BL) and directly generate metadata for existing video.
For Profile 5, encoding is necessary because there is no cross-compatibility.
In the early days of Dolby Vision, some content skipped the encoding step, such as Profile 4.2 (SDR Compatible, BL+EL+RPU) and Profile 8.2 (SDR Compatible, BL+RPU) both came from directly generating metadata, which can be efficiently "Upgrade" existing content. They have an automatically calculated reverse mapping, which is no longer recommended for professional content.

Profile 8.4 also falls into this category, except that they are usually user-generated content, mainly sourced from Apple device recordings or post-editing encodings.
Indeed, some Android devices also support recording, from Xiaomi and Vivo. Encoding is also supported in DaVinci Resolve for Windows (available with a professional license, or running on a certified Windows device, or connected to a certified monitor).

For user-generated Profile 8.4 content, per-frame metadata related to luminance (L1, L3, L4) is dynamically calculated along with HLG encoding. They both have static (preset) composer metadata for *approximately* converting HLG to PQ. In this way, display management is no different from PQ-based Profiles. In fact, the Dolby Vision pipeline does not need to know the characteristics of BL, with help of the composer. However, there are differences in the implementation of many software or device firmware. Some (including Apple) do not use composer metadata, but use other more precise methods. Some implementations may convert the super white region of the HLG to a signal over 1000nits, while others ignore this part.

The content may also contain static L8 Trim metadata, which is seen on iPhones running iOS 15 or later, Apple's "second version" of Profile 8.4. I guess this may be a preset made by Apple to better match the mapping effect with other content. Yes, Trim also works with Profile 8.4, but Dolby or Apple haven't made manual adjustments available to users (even pros).

By the way, the tone mapping on Apple devices is quite complicated, and there are many factors that can affect the mapping results, including the current brightness setting, ambient brightness, True Tone mode, low power mode, etc., not just by Dolby Vision display management decisions.

**Comment by joaoserra** [comment]
Thanks @saindriches and @quietvoid !

There's a lot to unpack here and I just went down the rabbit hole even deeper...
I already did know Dolby Vision and its multiple profiles and variants was a crazy fragmented/complex format, but it's even more fragmented than I thought.

So, i've read the documents referenced by @quietvoid  and found out there's actually two sets of metadata: the composer metadata and the actual HDR metadata(that closely resemble what i knew from ST 2094-10 - the RPU block syntax was defined before ST 2094, that's why they don't map 100%, but they're mostly the same thing.)
It's now clear to me the purpose of the composer metadata - Inverse mapping the EL layer(and the BL layer if it's SDR like in profile 4.2 and 8.2) and "join" the BL+EL in a single image.
This compose step it's actually done in the source device (i.e. BluRay player) and that's why the "official"(aka Supported by HDMI Forum) Dolby Vision support in HDMI 2.1 only standardizes the HDR metadata in CEA-861(basically a copy of ST 2094) and does not say anything about the composer metadata (it's not carried by HDMI, so it's not relevant)
Did i say anything incorrect so far?

I also did know that profile 5 used the ICtCp/ITP colorspace, but I thought this was a profile 5 specific thing and not used on other profiles.	 

So, I've looked at the metadata from profile 8.1 and also found composer metadata, I was only expecting to find HDR Metadata(ST 2094-10). Being a recent profile I was not even expecting to find the proprietary RPU block with both composer+hdr metadata in NALU 62. My expectation was only to find the standard ST 2094-10 block in a HEVC SEI message (and not the not standard HEVC NAL 62 RPU block) 
The whole idea of ST 2094 is to have a single HDR10 base layer plus multiple metadata types. (i.e. One single HDR10 image plus, say DV (ST 2094-10) and HDR10+(ST 2094-40) on a single file.
Now, what I fail to understand is the purpose of the composer metadata in profile 8.1. There's no need to do any reverse mapping or dual-layer processing as 8.1 is already single-layer with a HDR10 BL. 

@saindriches,  you mentioned that profile 8.4 has a static(when you say static, i guess it's the same for every iPhone file) composer metadata block to help the DV engine convert HLG to PQ. Does HLG's standard BT.2020 YCbCr also gets converted to ICtCp in the process? If yes, that made me thinking: Maybe that's the purpose of composer metadata in profile 8.1? 
If every input get's converted to ICtCp, like you said, that makes the work of the DV Engine pipeline more simple. And I guess that conversion to ICtCp requires composer metadata and that's why all profiles have it.

**Reply by quietvoid** [reply]
>Now, what I fail to understand is the purpose of the composer metadata in profile 8.1. There's no need to do any reverse mapping or dual-layer processing as 8.1 is already single-layer with a HDR10 BL.

For profile 8.1 you'd notice that while there is mapping metadata (for the composer), it is basically no-op. No reshaping.
Profile 5 requires reshaping, and 8.4 also has the static reshaping, at least the ones from iPhone.

I think it's essentially to keep backwards compatibility, at least in parts.

**Reply by joaoserra** [reply]
Thanks! Composer metadata on profile 8.1 is indeed only default values. Like you said it's a no op.

About 8.4, what you mean by static? The composer metadata doesn't change at all? (at least for the same iOS version) I've checked two different 8.4 samples of the older CMv2.9 with only L1 and indeed the composer metadata is the same in both files.(Only L1 data changes) However, i also checked a 8.4 file from a newer version of iOS with CMv4.0 present and the composer metadata is different from CMv2.9 (lot's of coefficients are different). 
Maybe Apple has a set of composer metadata for CMv2.9 and another for CMv4.0 files? (I don't have a second 8.4 test file with CMv4.0)

How about the ICtCp/ITP colorspace? Is this a profile 5 specific thing or the composer reshapes everything to ICtCp before sending the data to the HDMI interface?

You mention that profile 5 is also reshaped? What exactly is the composer reshaping on profile 5?

**Reply by quietvoid** [reply]
> About 8.4, what you mean by static? The composer metadata doesn't change at all?

Yes, the reshaping curves are always the same. As it's a static mapping from HLG to PQ.
Dolby has the means to do it dynamically but clearly it's not documented: https://github.com/DolbyLaboratories/dolby-vision-contents/issues/1

> However, i also checked a 8.4 file from a newer version of iOS with CMv4.0 present and the composer metadata is different from CMv2.9 (lot's of coefficients are different)

The coefficients have changed from OS updates.

> You mention that profile 5 is also reshaped? What exactly is the composer reshaping on profile 5?

Usually only luma is reshaped on profile 5, but nothing stops anyone from reshaping chroma as well (Apple does this).
Apple also has a static luma+chroma reshaping for their profile 5 encodes.

> How about the ICtCp/ITP colorspace? Is this a profile 5 specific thing or the composer reshapes everything to ICtCp before sending the data to the HDMI interface?

IPT is both a profile 5 thing and used internally by the tone mapping engine.
I'm not exactly sure of the details on the image sent to HDMI, as the DM metadata is the one that contains the output system coefficients.

**Reply by joaoserra** [reply]
Thanks!

> Usually only luma is reshaped on profile 5, but nothing stops anyone from reshaping chroma as well (Apple does this).
> Apple also has a static luma+chroma reshaping for their profile 5 encodes.

But what is the purpose exactly of that reshaping? What exactly is the composer trying to achieve in profile 5?
Maybe I'm not fully understanding the purpose of the composer.
My understanding is that the composer does two things:
    1) Map the BL EOTF to PQ. (i.e. map SDR or HLG to PQ)
    2) inverse quantize the EL and "join" the BL+EL in order to generate the final image.

Profile 5 is single layer, so the composer only processes the BL (it only uses the mapping/prediction part - Figure 2 of ETSI GS CCM 001)
ETSI GS CCM 001 specifies that the BL input format must be a PQ or a SDR EOTF signal (and I guess also HLG EOTF, but ETSI GS CCM 001 is older than profile 8.4) and the output is PQ signal.
But profile 5 is already a PQ signal, the only difference is that it uses the ICtCp colorspace, so what exactly is the composer trying to achieve in profile 5?

> IPT is both a profile 5 thing and used internally by the tone mapping engine.
> I'm not exactly sure of the details on the image sent to HDMI, as the DM metadata is the one that contains the output system coefficients.

After reading https://patents.google.com/patent/US20150248747A1/en it's now clear to me that the tone mapping engine works internally in ICtCp colorspace.
That patent says that whatever color space enters the pipeline will be converted to ICtCp. 
ETSI GS CCM says that the composer supports either YCbCr or ICtCp as input and also can output YCbCr or ICtCp. 
I know you said you're not sure of the exact details of the HDMI transmission, but do you know if the composer does any type of color space conversion? Or color space conversion is only done by the tone mapping engine in the sink device and the composer uses whatever colorspace the BL uses.

Btw, do you know any relevant patents/documents about the DV HDMI transmission format?

**Reply by quietvoid** [reply]
>But profile 5 is already a PQ signal, the only difference is that it uses the ICtCp colorspace.

This isn't the case, the source is PQ but it is reshaped to something close to 2.2 gamma for encoding in profile 5.
This is supposedly done to improve compression.

This reshaping process generates metadata for the reverse operation, and that's what is saved in the RPU.
On playback it has to be reshaped backwards back to PQ, using the metadata.

> Btw, do you know any relevant patents/documents about the DV HDMI transmission format?

There was previous discussion here: https://github.com/quietvoid/dovi_tool/discussions/48
HDMI transmission was already implemented in this project: https://www.avsforum.com/threads/dedicated-raspberry-pi-pgenerator-thread-set-up-configuration-updates-special-features-general-usage-tips.3167475/

It implies driver patches and libdrm code. If you look close enough you can find the source code.
However I don't think it's fit for actual playback, as the Dolby Vision signal has extra unwanted black pixels.

If you want more info you could look up info on LLDV, which is essentially the source device doing all the work.
There are knowledgeable people on AVSForum, especially the [HDFury threads](https://www.avsforum.com/threads/dolby-vision-including-hdr10-conversion-w-dtm-on-projectors.3097934).

**Comment by joaoserra** [comment]
> This isn't the case, the source is PQ but it is reshaped to something close to 2.2 gamma for encoding in profile 5. This is supposedly done to improve compression.
> 
> This reshaping process generates metadata for the reverse operation, and that's what is saved in the RPU. On playback it has to be reshaped backwards back to PQ, using the metadata.

Can you refer me to whatever patents/documents that describes that 2.2 gamma?

**Reply by quietvoid** [reply]
I just pulled the 2.2 out of nowhere, I don't think it's specified at all.
The reshaping can make it look however.
It just looks like regular SDR, not PQ.

**Comment by joaoserra** [comment]
Thanks for all the information!

Just two more questions:

- Have you actually seen standards compliant ST 2094-10 DV files in the wild? (i.e. without RPU in NAL 62 and only ST 2094-10 in SEI messages? I'm starting to get the impression that ST 2094-10 only exists "on paper" and was only done to satisfy the DVB/ATSC standard bodies...
- How does a dual DolbyVision/HDR10+ file on Blu-Ray look like? Standard DV profile 7 RPU in NAL 62 plus HDR10+/ST 2094-40 in SEI messages?

**Comment by quietvoid** [comment]
The only 2094-10 sample I'm aware of is from https://dash.akamaized.net/dash264/TestCasesMCA/dolby/6
It follows the structure defined here: https://dashif-documents.azurewebsites.net/DASH-IF-IOP/master/DASH-IF-IOP.html#codecs-dolbyvision-enhancement-layer

> How does a dual DolbyVision/HDR10+ file on Blu-Ray look like? Standard DV profile 7 RPU in NAL 62 plus HDR10+/ST 2094-40 in SEI messages?

The ST 2094-40 SEI are in the base layer, while the EL is a second stream, in which the RPU is in NALUs with type 62.

**Reply by joaoserra** [reply]
> The only 2094-10 sample I'm aware of is from https://dash.akamaized.net/dash264/TestCasesMCA/dolby/6
> It follows the structure defined here: https://dashif-documents.azurewebsites.net/DASH-IF-IOP/master/DASH-IF-IOP.html#codecs-dolbyvision-enhancement-layer

Thanks! Will test it! Have you tried to playback this on actual DV players?

**Reply by quietvoid** [reply]
Nope, never tested muxing it into a single file.
